app/services/mentee_application_transition_service.rb

```diff
@@ -35,34 +35,28 @@ class InvalidTransitionError < StandardError; end
     }
   }.freeze
 
-  def promote!(application:, reviewer:, note: nil)
+  def call(application:, reviewer:, action:, note: nil)
     status = application.current_status.to_sym
-    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? :promote
-    promote_transition = STATUS_TRANSITION_MAPPING[status][:promote_transition]
-    application.mentee_application_states.create!(status: promote_transition, reviewer:, note:)
-    invoke_side_effects(application, promote_transition, reviewer)
+    # enforce that the action is cast to a symbol before performing guard clause
+    action = action.to_sym
+    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? action
+    transition_status = action == :reject ? :rejected : STATUS_TRANSITION_MAPPING[status][:promote_transition]
+    application.mentee_application_states.create!(status: transition_status, reviewer:, note:)
+    # handle side effects
+    case transition_status
+    when :coding_challenge_sent then code_challenge_side_effects(application)
+    when :accepted then accepted_side_effects(application)
+    when :rejected then rejected_side_effects(application)
+    end
     true
   end
 
-  def reject!(application:, reviewer:, note: nil)
-    status = application.current_status.to_sym
-    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? :reject
-    promote_transition = :rejected
-    application.mentee_application_states.create!(status: promote_transition, reviewer:, note:)
-    invoke_side_effects(application, promote_transition, reviewer)
-    true
+  def valid_transitions(status:)
```

Comment:
I want to define an interface to get valid transitions to enforce message passing, rather than allowing an application to reach into the internal state of the service object.